aten/src/ATen/cuda/cub.cu

```diff
@@ -0,0 +1,151 @@
+                keys_in_, keys_out_, n,
+                begin_bit, end_bit, c10::cuda::getCurrentCUDAStream());
+  } else {
+    CUB_WRAPPER(NO_ROCM(at_cuda_detail)::cub::DeviceRadixSort::SortKeys,
+                keys_in_, keys_out_, n,
+                begin_bit, end_bit, c10::cuda::getCurrentCUDAStream());
+  }
+}
+
+template<typename scalar_t>
+void unique(const scalar_t *input, scalar_t *output, int64_t *num_selected_out, int64_t num_items) {
+  TORCH_CHECK(num_items <= std::numeric_limits<int>::max(),
+              "cub unique does not support more than INT_MAX elements");
+  CUB_WRAPPER(NO_ROCM(at_cuda_detail)::cub::DeviceSelect::Unique,
+              input, output, num_selected_out, num_items, at::cuda::getCurrentCUDAStream());
+}
+
+template <typename scalar_t>
+void run_length_encode(const scalar_t *input, scalar_t *output, int64_t *counts_out,
+                       int64_t *length_out, int64_t num_items) {
+  TORCH_CHECK(num_items <= std::numeric_limits<int>::max(),
+              "cub run_length_encode does not support more than INT_MAX elements");
+  CUB_WRAPPER(
+      NO_ROCM(at_cuda_detail)::cub::DeviceRunLengthEncode::Encode,
+      input, output, counts_out, length_out, num_items,
+      at::cuda::getCurrentCUDAStream());
+}
+
+#define AT_INSTATIATE_CUB_TEMPLATES(scalar_t, ScalarType)           \
+  template void radix_sort_keys(                                    \
+      const scalar_t *keys_in, scalar_t *keys_out, int64_t n,       \
+      bool descending, int64_t begin_bit, int64_t end_bit);         \
+  template void unique(                                             \
+      const scalar_t *input, scalar_t *output,                      \
+      int64_t *num_selected_out, int64_t num_items);                \
+  template void run_length_encode(                                  \
+      const scalar_t *input, scalar_t *output, int64_t *counts_out, \
+      int64_t *length_out, int64_t n);
+
+AT_FORALL_SCALAR_TYPES_AND2(Bool, Half, AT_INSTATIATE_CUB_TEMPLATES)
+
+namespace {
+template <typename scalar_t>
+struct SumOp {
+  __device__ scalar_t operator () (scalar_t a, scalar_t b) const {
+    return a + b;
+  }
+};
+}
+
+template <typename input_t, typename output_t>
+void inclusive_sum(const input_t *input, output_t *output, int64_t num_items) {
+  using scalar_t = std::common_type_t<input_t, output_t>;
+  inclusive_scan(input, output, SumOp<scalar_t>{}, num_items);
+}
+
+template void inclusive_sum(const int32_t *input, int32_t *output, int64_t num_items);
+template void inclusive_sum(const int64_t *input, int64_t *output, int64_t num_items);
+template void inclusive_sum(const int32_t *input, int64_t *output, int64_t num_items);
+
+template <typename input_t, typename output_t>
+void exclusive_sum(const input_t *input, output_t *output, int64_t num_items) {
+  using scalar_t = std::common_type_t<input_t, output_t>;
+  exclusive_scan(input, output, SumOp<scalar_t>{}, scalar_t(0), num_items);
+}
```

Comment:
This behavior is different from `cub::DeviceScan::ExclusiveSum` (see https://github.com/pytorch/pytorch/pull/67650/files#r745111950), it's better to document this behavior some way, for example, renaming this function as `exclusive_sum_in_common_type`

Reply:
Done, although I've left `exclusive_sum` as an alias that requires the input and output types to be the same.